pkg/apis/kops/cluster.go

```diff
@@ -191,6 +191,8 @@ type ClusterSpec struct {
 	SysctlParameters []string `json:"sysctlParameters,omitempty"`
 	// RollingUpdate defines the default rolling-update settings for instance groups
 	RollingUpdate *RollingUpdate `json:"rollingUpdate,omitempty"`
+	// ServiceOIDCProvider defines the OIDC provider setup for the cluster (AWS only)
+	ServiceOIDCProvider *ServiceOIDCProviderSpec `json:"serviceOIDCProvider,omitempty"`
```

Comment:
Yeah - naming for this is going to be tough.

* It's not technically specific to AWS; if we publish the JWKS doc from e.g. an on-metal cluster we should be able to authenticate to AWS.
* Similarly GCE has some support for this now also, in the form of GKE workload identity, so it's not just _to_ AWS.
* It's actually even broader than this; we can also authenticate service accounts across clusters etc.

The word that keeps sticking in my mind is "discovery", and this is specifically "identity discovery".  I think it's also possible to do cross-cluster discovery using that same bucket (and those files wouldn't necessarily have to be public, either)

Maybe something like

```
discovery:
  store: s3://some-other-bucket/clusters
  identity: true
```

I think technically we could even reuse the same S3 bucket, I can't imagine it's worth it from an isolation point of view though :-)

Reply:
I don't know how things other than AWS consume OIDC, but I would hope the `IssuerCAThumbprints` field wouldn't be copied by other cloud providers. It would be a lot more sane if one could just configure AWS with the identity document directly, rather than faffing about with serving it on the public Internet. If AWS ever changes the CA they use for their S3 buckets, everyone's IRSA will likely break immediately.

Reply:
It's a good point on the rotations - I presume this means that S3 effectively can't change CAs quickly, though I don't know what happens in 2025 when the current cert expires, presumably some pre-publication.  (I think the cert fingerprint includes the expiration date?)

@rifelpet and I chatted on Friday, and we had an idea about an approach that might be a nice way to get this merged: we define a feature flag (or two) that effectively does all this automagically in a (very) limited set of circumstances - currently you must use a public ELB, with a real DNS name, likely others we don't even know about.  The featureflag enables the apiserver with anonymous-access turned on (which normally we don't like to do in kops, particularly for public ELBs), we also grant the anonymous-access user permission to read the JWKS documents directly.

We should need to find a way to not encourage this as a real scenario (e.g. I don't want to preclude private ELBs).  But it means we don't need an API surface in this PR and we don't need kops-controller to publish to an S3 bucket in the one I added.  That PR is https://github.com/kubernetes/kops/pull/9352

We can then discuss e.g. whether to have a public bucket, how to name the clusters, what the API should be for enabling AWS integration in a more lesiurely way.  But we don't block working on e.g. per-pod roles or figuring out different ways to publish the JWKS doc (e.g. publish from kops CLI vs publish from kops-controller).

This also means that upstream (k/k) can start to see some real test coverage of the feature, and get it to GA.